File: edk/examples/huestream_example_gui_win/BrightnessSawEffect.cpp
/*******************************************************************************
 Copyright (C) 2018 Philips Lighting Holding B.V.
 All Rights Reserved.
 ********************************************************************************/

#include <huestream/effect/animation/animations/ConstantAnimation.h>
#include <huestream/effect/animation/animations/SequenceAnimation.h>
#include <huestream/effect/animation/animations/TweenAnimation.h>
#include <huestream_example_gui_win/BrightnessSawEffect.h>

#include <string>
#include <memory>

using std::string;

namespace huestream {

    BrightnessSawEffect::BrightnessSawEffect(const string &name, unsigned int layer) : AreaEffect(name, layer) {}

    void BrightnessSawEffect::PrepareEffect(std::shared_ptr<ConstantAnimation> speedAnimation,
                                            std::shared_ptr<ConstantAnimation> alphaAnimation) {
        auto r = std::make_shared<SequenceAnimation>(INF);

        auto rampUp = std::make_shared<TweenAnimation>(0, 1, 4000, TweenType::Linear);
        r->Append(rampUp, "Up");

        auto rampDown = std::make_shared<TweenAnimation>(1, 0, 4000, TweenType::Linear);
        r->Append(rampDown, "Down");

        auto g = std::make_shared<ConstantAnimation>(0);
        auto b = std::make_shared<ConstantAnimation>(0);

        SetColorAnimation(r, g, b);
        SetOpacityAnimation(alphaAnimation);
        SetSpeedAnimation(speedAnimation);

        AddArea(Area::FrontHalf);
    }

}  // namespace huestream

File: edk/wrappers/csharp/huestream_csharp_example_wpf/BrightnessSawEffect.cs
using huestream;

namespace HueStreamExampleWPF
{
    class BrightnessSawEffect : AreaEffect
    {
        public BrightnessSawEffect(string name, uint priority) : base(name, priority)
        {
        }

        public void PrepareEffect()
        {
            var r =  new SequenceAnimation(double.MaxValue);

            var rampUp = new TweenAnimation(0, 1, 1000, TweenType.EaseInOutQuad);
            r.Append(rampUp, "Up");

            var rampDown = new TweenAnimation(1, 0, 1000, TweenType.EaseInOutQuad);
            r.Append(rampDown, "Down");

            var g = new ConstantAnimation(0);
            var b = new ConstantAnimation(0);

            SetColorAnimation(r, g, b);

            AddArea(Area.All);
        }
    }
}